### PR TITLE
prov/tcp: fix incorrect usage of av insert apis when multiplexing

### PR DESCRIPTION
**Pull request by ooststep**
When multiplexing av inserts to sub avs, correctly track and check the subav fi_addrs

### CODE REVIEW

File: prov/tcp/src/xnet_av.c [hunk 1]
```diff
@@ -97,23 +99,26 @@ static int xnet_mplex_av_insertsym(struct fid_av *av_fid, const char *node,
...
-				      &sub_fi_addr, flags, context);
-		if (ret)
+				      sub_fi_addr, flags, context);
+		if (ret <= 0)
 			break;
-		assert(*fi_addr == sub_fi_addr);
+		assert(!fi_addr || memcmp(fi_addr, sub_fi_addr,
+					  sizeof(fi_addr_t) * ret) == 0);
 	}
+	free(sub_fi_addr);
```
Comment (j-xiong) [inline prov/tcp/src/xnet_av.c:121]: The error cases are not handled properly. If `ofi_ip_av_insertsym` failed to insert all the addresses (i.e. `ret < nodecnt * svccnt`), `sub_fi_addr` would not have enough space to hold all the addresses returned from `fi_av_insertsym`. In addition, the final return value from this function may not conforming to the API definition.

File: prov/tcp/src/xnet_av.c [hunk 2]
```diff
@@ -97,26 +99,27 @@ static int xnet_mplex_av_insertsym(struct fid_av *av_fid, const char *node,
...
+	if (ret < nodecnt*svccnt)
 		goto out;
 	dlist_foreach_container(&av->subav_list, struct fid_list_entry,	item, entry) {
 		subav_fid = container_of(item->fid, struct fid_av, fid);
 		ret = fi_av_insertsym(subav_fid, node, nodecnt, service, svccnt,
-				      &sub_fi_addr, flags, context);
-		if (ret)
+				      sub_fi_addr, flags, context);
+		if (ret <= nodecnt*svccnt)
 			break;
```
Comment (j-xiong) [inline prov/tcp/src/xnet_av.c:116]: What is the expectation of the resulting AV when some of the addresses fail to insert?  If the function returns a value < nodecnt * svccnt due to L109 or L115, the state of the main AV and sub AVs are going to be inconsistent. Will the provider be able to handle the case that the application proceeds to use the AV for one of the addressed successfully inserted in the partial successful call?  This is not an issue specifically for this PR, but rather is related to the suv-AV approach.

File: prov/tcp/src/xnet_av.c [hunk 3]
```diff
@@ -69,23 +69,25 @@ static int xnet_mplex_av_insert(struct fid_av *av_fid, const void *addr, size_t
 	int ret;
 	struct fid_list_entry *item;
 	struct fid_av *subav_fid;
-	fi_addr_t sub_fi_addr;
+	fi_addr_t *sub_fi_addr;
 	struct xnet_mplex_av *av = container_of(av_fid, struct xnet_mplex_av,
-						util_av.av_fid.fid);
-
+						util_av.av_fid);
+	sub_fi_addr = calloc(count, sizeof(fi_addr_t));
```
[the remark this answers is not in the document]
Comment (ooststep) [reply prov/tcp/src/xnet_av.c:75]: oops

File: prov/tcp/src/xnet_av.c (hunk 2 above)
Comment (ooststep) [reply prov/tcp/src/xnet_av.c:116]: Yeah, this is a fairly hard problem to handle I think.  If the av and sub-av are out of sync, it's basically a fatal error.  I suppose if the main av or sub-av fail to fully insert, we may need to roll back and fail the whole thing (for fi_av_insert, too).

File: prov/tcp/src/xnet_av.c [hunk 4]
```diff
@@ -97,26 +101,29 @@ static int xnet_mplex_av_insertsym(struct fid_av *av_fid, const char *node,
...
 						util_av.av_fid.fid);
-
+	sub_fi_addr = calloc(nodecnt*svccnt, sizeof(fi_addr_t));
+	if (!sub_fi_addr)
+		return -FI_ENOMEM;
 	ofi_genlock_lock(&av->lock);
 	ret = ofi_ip_av_insertsym(&av->util_av.av_fid, node, nodecnt,
 				  service, svccnt, fi_addr, flags, context);
-	if (ret)
+	if (ret < nodecnt*svccnt)
```
Comment (aingerson) [inline prov/tcp/src/xnet_av.c:113]: spacing
` ret < nodecnt * svccnt`

File: prov/tcp/src/xnet_av.c [hunk 5]
```diff
@@ -97,26 +101,29 @@ static int xnet_mplex_av_insertsym(struct fid_av *av_fid, const char *node,
...
 		goto out;
 	dlist_foreach_container(&av->subav_list, struct fid_list_entry,	item, entry) {
 		subav_fid = container_of(item->fid, struct fid_av, fid);
 		ret = fi_av_insertsym(subav_fid, node, nodecnt, service, svccnt,
-				      &sub_fi_addr, flags, context);
-		if (ret)
+				      sub_fi_addr, flags, context);
+		if (ret <= nodecnt*svccnt)
```
Comment (aingerson) [inline prov/tcp/src/xnet_av.c:119]: same here

File: prov/tcp/src/xnet_av.c [hunk 6]
```diff
@@ -97,26 +101,29 @@ static int xnet_mplex_av_insertsym(struct fid_av *av_fid, const char *node,
...
 			break;
-		assert(*fi_addr == sub_fi_addr);
+		assert(!fi_addr || memcmp(fi_addr, sub_fi_addr,
+					  sizeof(fi_addr_t) * nodecnt*svccnt) == 0);
```
Comment (aingerson) [inline prov/tcp/src/xnet_av.c:122]: and here

**Review by aingerson — COMMENTED** [review]
Just pointing out some missing space but otherwise looks good!

**Comment by ooststep** [comment]
@j-xiong when this is ready to merge, does it also need to be cherry picked or will future 2.0 releases come off master?

**Comment by j-xiong** [comment]
The main branch is for 2.0.

**Review by j-xiong — APPROVED** [review]
I'll leave the proper handling of partial success for future improvement.

**Pull request merged**